### PR TITLE
KNOX-2126. Add an API to retrieve the public key used for Knox JWT verification.

### PR DESCRIPTION
**Pull request by sidseth**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

## How was this patch tested?
Not tested yet. Looking for some feedback on client side changes, if any are required.

### CODE REVIEW

File: gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java [hunk 1]
```diff
@@ -278,6 +284,39 @@ public Response revoke(String token) {
...
+
+      String jsonResponse = JsonUtils.renderAsJsonString(map);
+      return Response.ok().entity(jsonResponse).build();
+    } catch (KeyStoreException | KeystoreServiceException e) {
+      return Response.serverError().build();
+      // TODO: Look at serverError(). ... (Ref: ServiceDefinitionResource for more info)
+    }
+  }
+
+  private String getSigningKeyAlias() {
```
Comment (risdenk) [inline gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java]: I think the code already accounts for the `alias == null` case?

https://github.com/apache/knox/blame/master/gateway-server/src/main/java/org/apache/knox/gateway/config/impl/GatewayConfigImpl.java#L749

File: gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java [hunk 2]
```diff
@@ -278,6 +284,39 @@ public Response revoke(String token) {
...
+    KeystoreService ks = services.getService(ServiceType.KEYSTORE_SERVICE);
+
+    try {
+      PublicKey publicKey = ks.getSigningKeystore().getCertificate(getSigningKeyAlias()).getPublicKey();
+      publicKey.getAlgorithm();
+      byte[] pubKeyBytes = publicKey.getEncoded();
+      String base64Key = Base64.encodeBase64String(pubKeyBytes);
+      Map<String, String> map = new HashMap<>();
+      map.put(PUBLIC_KEY, base64Key);
+      map.put(ALGORITHM, publicKey.getAlgorithm());
```
Comment (risdenk) [inline gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java]: Do we need to expose the algorithm here? Wouldn't it be better to normalize? Thinking about consumers of this they would need to be able to deal with different algorithms before reading the public key?

File: gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java [hunk 3]
```diff
@@ -278,6 +284,39 @@ public Response revoke(String token) {
...
+    GatewayServices services = (GatewayServices) request.getServletContext()
+        .getAttribute(GatewayServices.GATEWAY_SERVICES_ATTRIBUTE);
+
+    KeystoreService ks = services.getService(ServiceType.KEYSTORE_SERVICE);
+
+    try {
+      PublicKey publicKey = ks.getSigningKeystore().getCertificate(getSigningKeyAlias()).getPublicKey();
+      publicKey.getAlgorithm();
+      byte[] pubKeyBytes = publicKey.getEncoded();
+      String base64Key = Base64.encodeBase64String(pubKeyBytes);
```
Comment (risdenk) [inline gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java]: Is this base64 encoded pem format then? Might be worth a separate method that takes the `PublicKey` and describes exactly what is being returned in what format. PEM/DER/base64/etc. 

I don't think it changes the functionality but would make it more clear.

File: gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java [hunk 4]
```diff
@@ -278,6 +284,39 @@ public Response revoke(String token) {
...
+  @Produces ({APPLICATION_JSON})
+  public Response publicKey() {
+    GatewayServices services = (GatewayServices) request.getServletContext()
+        .getAttribute(GatewayServices.GATEWAY_SERVICES_ATTRIBUTE);
+
+    KeystoreService ks = services.getService(ServiceType.KEYSTORE_SERVICE);
+
+    try {
+      PublicKey publicKey = ks.getSigningKeystore().getCertificate(getSigningKeyAlias()).getPublicKey();
+      publicKey.getAlgorithm();
```
Comment (risdenk) [inline gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java]: Nit: looks like an extra line?

Comment (pzampino) [inline gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java]: The result of this (publicKey.getAlgorithm()) is ignored? Is it necessary?

File: gateway-service-knoxtoken/src/test/java/org/apache/knox/gateway/service/knoxtoken/TokenServiceResourceTest.java [hunk 5]
```diff
@@ -778,6 +785,32 @@ public void testTokenRevocation_Enabled_WithRenewersWithValidSubject() throws Ex
...
+        Base64.decodeBase64(responseMap.get(TokenResource.PUBLIC_KEY)), publicKey.getEncoded());
+
+  }
+
+  @Test
+  public void testGetPublicKey_error() throws Exception {
+    TokenResource tr = createTokenResourceForPublicKeyTests(true);
+
+    Response response = tr.publicKey();
+    assertEquals("Unexpoected response code", Status.INTERNAL_SERVER_ERROR.getStatusCode(),
```
Comment (pzampino) [inline gateway-service-knoxtoken/src/test/java/org/apache/knox/gateway/service/knoxtoken/TokenServiceResourceTest.java]: Typo: Unexpoected

File: gateway-service-knoxtoken/src/test/java/org/apache/knox/gateway/service/knoxtoken/TokenServiceResourceTest.java [hunk 6]
```diff
@@ -778,6 +785,32 @@ public void testTokenRevocation_Enabled_WithRenewersWithValidSubject() throws Ex
...
+  public void testGetPublicKey() throws Exception {
+    TokenResource tr = createTokenResourceForPublicKeyTests(false);
+
+    Response response = tr.publicKey();
+    assertEquals("Unexpected response code", Status.OK.getStatusCode(), response.getStatus());
+    String retString = response.getEntity().toString();
+    Map<String, String> responseMap = parseJSONResponse(retString);
+    assertNotNull("PublicKey is null", responseMap.get(TokenResource.PUBLIC_KEY));
+    assertNotNull("Algorithm is null", responseMap.get(TokenResource.ALGORITHM));
+    assertEquals("Mismatched algorithm", responseMap.get(TokenResource.ALGORITHM),
```
Comment (pzampino) [inline gateway-service-knoxtoken/src/test/java/org/apache/knox/gateway/service/knoxtoken/TokenServiceResourceTest.java]: Nitpick: Generally, the expected value is first in assertEquals(expected, actual)

File: gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java [hunk 7]
```diff
@@ -278,6 +284,39 @@ public Response revoke(String token) {
...
+      byte[] pubKeyBytes = publicKey.getEncoded();
+      String base64Key = Base64.encodeBase64String(pubKeyBytes);
+      Map<String, String> map = new HashMap<>();
+      map.put(PUBLIC_KEY, base64Key);
+      map.put(ALGORITHM, publicKey.getAlgorithm());
+
+      String jsonResponse = JsonUtils.renderAsJsonString(map);
+      return Response.ok().entity(jsonResponse).build();
+    } catch (KeyStoreException | KeystoreServiceException e) {
+      return Response.serverError().build();
```
Comment (pzampino) [inline gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java]: We should at least log the exception message here in the gateway log.
I'm always conflicted about how much detail we should return to the client in such cases.

File: gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java (hunk 1 above)
Comment (sidseth) [reply gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java]: Removed the check.

File: gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java (hunk 2 above)
Comment (sidseth) [reply gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java]: Good question. Let me evaluate this a little more, and get back.

File: gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java (hunk 4 above)
Comment (sidseth) [reply gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java]: Removed. Left over from testing.

Comment (sidseth) [reply gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java]: Removed. Left over from testing.

File: gateway-service-knoxtoken/src/test/java/org/apache/knox/gateway/service/knoxtoken/TokenServiceResourceTest.java (hunk 5 above)
Comment (sidseth) [reply gateway-service-knoxtoken/src/test/java/org/apache/knox/gateway/service/knoxtoken/TokenServiceResourceTest.java]: Fixed

File: gateway-service-knoxtoken/src/test/java/org/apache/knox/gateway/service/knoxtoken/TokenServiceResourceTest.java (hunk 6 above)
Comment (sidseth) [reply gateway-service-knoxtoken/src/test/java/org/apache/knox/gateway/service/knoxtoken/TokenServiceResourceTest.java]: Won't count that as a nitpick. Expected should go first per the API. Fixed.

File: gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java (hunk 7 above)
Comment (sidseth) [reply gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java]: Will add a log line.
Yup - what level of information goes back to the client is always problematic. In this particular case, I doubt the client learns much from the error going back (There's no parameters etc for the client to get wrong which they could try fixing), and this requires a server side fix.

File: gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java (hunk 3 above)
Comment (sidseth) [reply gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java]: @risdenk - could you please elaborate a bit. Are you also suggesting including this information in the response (i.e. will the client need this to interpret the response correctly - that is why I had initially included the algorithm. I'd imagine the format is actually more important).

File: gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java (hunk 7 above)
Comment (pzampino) [reply gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java]: I agree with that assessment of the response content.

**Comment by risdenk** [comment]
@sidseth  what is the status of this change?

**Comment by smolnar82** [comment]
This PR has not been touched for over a year ago; closing it.

**Pull request closed without merging**